crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    context_data::db_data_provider::diesel_macro::read_only_blocking, error::Error,
+    types::digest::Digest,
+};
+use diesel::{
+    r2d2::{self, ConnectionManager},
+    ExpressionMethods, OptionalExtension, PgConnection, QueryDsl, RunQueryDsl,
+};
+use move_bytecode_utils::module_cache::SyncModuleCache;
+use std::{env, str::FromStr, sync::Arc};
+use sui_indexer::{
+    models_v2::{
+        checkpoints::StoredCheckpoint, epoch::StoredEpochInfo, transactions::StoredTransaction,
+    },
+    schema_v2::{checkpoints, epochs, transactions},
+};
+pub type PgConnectionPool = diesel::r2d2::Pool<ConnectionManager<PgConnection>>;
+pub type PgPoolConnection = diesel::r2d2::PooledConnection<ConnectionManager<PgConnection>>;
+
+use super::module_resolver::PgModuleResolver;
+
+pub(crate) mod diesel_macro {
+    macro_rules! read_only_blocking {
+        ($pool:expr, $query:expr) => {{
+            let mut pg_pool_conn =
+                crate::context_data::db_data_provider::get_pg_pool_connection($pool)?;
+            pg_pool_conn
+                .build_transaction()
+                .read_only()
+                .run($query)
+                .map_err(|e| Error::Internal(e.to_string()))
+        }};
+    }
+    pub(crate) use read_only_blocking;
+}
+
+pub fn get_pg_pool_connection(pool: &PgConnectionPool) -> Result<PgPoolConnection, Error> {
+    pool.get().map_err(|e| {
+        Error::Internal(format!(
+            "Failed to get connection from PG connection pool with error: {:?}",
+            e
+        ))
+    })
+}
+
+pub fn establish_connection_pool() -> PgConnectionPool {
+    let database_url = env::var("PG_DB_URL").expect("PG_DB_URL must be set");
+    let manager = ConnectionManager::<PgConnection>::new(database_url);
+    r2d2::Pool::builder()
+        .build(manager)
+        .expect("Failed to create pool.")
+}
+
+#[allow(dead_code)]
+pub(crate) struct PgManager {
+    pub pool: PgConnectionPool,
+    pub module_cache: Arc<SyncModuleCache<PgModuleResolver>>,
+}
+
+impl PgManager {
+    pub(crate) fn new() -> Self {
+        let pool = establish_connection_pool();
+        Self {
+            pool: pool.clone(),
+            module_cache: Arc::new(SyncModuleCache::new(PgModuleResolver::new(pool))),
+        }
+    }
+
+    pub(crate) async fn fetch_tx(&self, digest: &str) -> Result<Option<StoredTransaction>, Error> {
+        let digest = Digest::from_str(digest)?;
+        read_only_blocking!(&self.pool, |conn| {
```

Comment:
Any and all reads from the DB are done in a "blocking" manner. This means that they cannot be run from inside a normal "async" function. If you look at the indexer code you'll see we use a spawn_blocking wrapper in order to call these blocking methods from async contexts